app/application/src/main/scala/myapp/application/projection/deposit/DepositSourceProvider.scala

```diff
@@ -0,0 +1,85 @@
+package myapp.application.projection.deposit
+
+import akka.NotUsed
+import akka.actor.typed.ActorSystem
+import akka.projection.scaladsl.SourceProvider
+import akka.stream.scaladsl.Source
+import lerna.log.AppLogging
+import lerna.util.lang.Equals._
+import myapp.readmodel.JDBCService
+import myapp.readmodel.schema.Tables
+import myapp.utility.tenant.AppTenant
+
+import scala.concurrent.Future
+
+private[deposit] class DepositSourceProvider(
+    jdbcService: JDBCService,
+    tables: Tables,
+    config: DepositProjectionConfig,
+)(implicit
+    system: ActorSystem[Nothing],
+    tenant: AppTenant,
+) extends SourceProvider[DepositProjection.Offset, Deposit]
+    with AppLogging {
+  import DepositSourceProvider._
+  import lerna.log.SystemComponentLogContext._
+  import tables._
+  import tables.profile.api._
+
+  override def source(offset: () => Future[Option[DepositProjection.Offset]]): Future[Source[Deposit, NotUsed]] = {
+    import system.executionContext
+    offset().map(createSource)
+  }
+
+  override def extractOffset(envelope: Deposit): DepositProjection.Offset = envelope.depositId.value
+
+  override def extractCreationTime(envelope: Deposit): Long = envelope.createdAt.toEpochMilli
+
+  private[this] type OffsetOption = Option[DepositProjection.Offset]
+
+  private[this] def createSource(initialOffset: OffsetOption): Source[Deposit, NotUsed] = {
+    logger.info(s"Source created (Offset: ${initialOffset.toString})")
+    Source
+      .unfoldAsync[(OffsetOption, PollingControl), Seq[DepositStoreRow]]((initialOffset, Continue)) {
+        case (offset, control) =>
+          import system.executionContext
+          def fetchNextBatch(): Future[Option[((OffsetOption, PollingControl), Seq[DepositStoreRow])]] = {
+            for {
+              deposits <- fetchDeposits(offset)
+            } yield {
+              // 結果が Offset (depositId) の昇順でソート済みのため、最後の行に最新の Offset があるとみなせる
+              val nextOffset: OffsetOption    = deposits.lastOption.map(_.depositId).orElse(offset)
+              val hasMoreEvents               = deposits.size === config.pollingBatchSize
+              val nextControl: PollingControl = if (hasMoreEvents) Continue else ContinueDelayed
+              Option(((nextOffset, nextControl), deposits))
+            }
+          }
+          control match {
+            case Continue        => fetchNextBatch()
+            case ContinueDelayed => akka.pattern.after(config.pollingInterval)(fetchNextBatch())
+          }
+      }
+      .mapConcat(identity)
+      .map { row =>
+        Deposit(DepositId(row.depositId), row.accountNo, BigInt(row.amount), row.createdAt.toInstant)
+      }
+  }
+
+  private[this] def fetchDeposits(offset: OffsetOption): Future[Seq[DepositStoreRow]] = {
+    val query = offset match {
+      case Some(offset) =>
+        DepositStore.filter(_.depositId > offset)
+      case None =>
+        DepositStore
```

Comment:
FYI

`filterOpt` が使えました。

```
DepositStore.filterOpt(offset)(_.depositId > _)
```

※ 内部的には `withFilter` が使われている点が違いますが、for式用のmethodなだけで 効果は `filter` と同じそうです。

```scala
  def filterOpt[V, T <: Rep[_] : CanBeQueryCondition](optValue: Option[V])(f: (E, V) => T): Query[E, U, C] =
    optValue.map(v => withFilter(a => f(a, v))).getOrElse(this)
```
```scala

  /** Select all elements of this query which satisfy a predicate. This method
    * is used when desugaring for-comprehensions over queries. There is no
    * reason to call it directly because it is the same as `filter`. */
  def withFilter[T : CanBeQueryCondition](f: E => T) = filterHelper(f, identity)
```

Reply:
@tksugimoto コメントを追記しておきました。初学者が理解しやすいようにコードは `filter` のままにしています。

f927414